bpf/standalone/txg.py

```diff
@@ -0,0 +1,378 @@
+
+#define SPA_SYNC_DIRTY_INDEX  0;
+
+BPF_HASH(sync_info_map, u32, spa_sync_info);
+BPF_ARRAY(sync_dirty_map, spa_sync_dirty, 1);
+BPF_PERF_OUTPUT(sync_events);
+
+/*
+ * Store the spa_t address if the pool name matches POOL,
+ * "domain0" by default. Should only have to do the string
+ * comparision once.
+ */
+
+static inline void pool_set(spa_sync_info *info, spa_t *spa)
+{
+    char comparand[sizeof(POOL)];
+    bpf_probe_read(&comparand, sizeof(comparand), spa->spa_name);
+    char compare[] = POOL;
+    for (int i = 0; i < sizeof(comparand); ++i)
+       if (compare[i] != comparand[i])
+           return;
+    info->spa = spa;
+}
+
+/*
+ * Get the domain0 spa object and store "start" time and
+ * "time since last sync".
+ */
+int spa_sync_entry(struct pt_regs *ctx, spa_t *spa, uint64_t txg)
+{
+    u32 tid = bpf_get_current_pid_tgid();
+    spa_sync_info *info = sync_info_map.lookup(&tid);
+    if (info == NULL) {
+       spa_sync_info spa_info = {};
+       sync_info_map.insert(&tid, &spa_info);
+       info = &spa_info;
+
+       int dirty_index = SPA_SYNC_DIRTY_INDEX;
+       spa_sync_dirty *spa_dirty;
+       spa_dirty = sync_dirty_map.lookup(&dirty_index);
+       if (spa_dirty != NULL) {
+           spa_dirty->dirty_addr = &spa->spa_dsl_pool->dp_dirty_total;
+           spa_dirty->dirty_max = spa->spa_dsl_pool->dp_dirty_total;
+       }
+    }
+
+    if (info->spa == 0)
+        pool_set(info, spa);
+
+    if (info->spa != spa)
+        return 0;
+
+    info->start = bpf_ktime_get_ns();
+    info->txg = txg;
+    /*
+     * We may not have a "stop" time yet, so just print a (kind of bogus)
+     * zero value for "time since last sync" if that's the case. This only
+     * affects the first txg we see -- after that "stop" will be set by the
+     * previous txg.
+     */
+    info->time_since_last_sync = info->stop != 0 ?
+        info->start - info->stop : 0;
+    info->pool_sync_enter = 0;
+    info->delay_entry_time = 0;
+    info->delay_max = 0;
+    return 0;
+}
+
+/*
+ * Collect data required to know the percentage of syncing time spent
+ * in pass 1.
+ */
+int dsl_pool_sync_entry(struct pt_regs *ctx, dsl_pool_t *dp)
+{
+    u32 tid = bpf_get_current_pid_tgid();
+    spa_sync_info *info = sync_info_map.lookup(&tid);
+    if (info == NULL || info->spa != dp->dp_spa ||
+        info->spa->spa_sync_pass != 1) {
+        return 0;
+    }
+
+    info->pool_sync_enter =  bpf_ktime_get_ns();
+    return 0;
+}
+
+int dsl_pool_sync_return(struct pt_regs *ctx)
+{
+    u32 tid = bpf_get_current_pid_tgid();
+    spa_sync_info *info = sync_info_map.lookup(&tid);
+    if (info == NULL || info->pool_sync_enter == 0) {
+        return 0;
+    }
+
+    info->elapsed_pass1 = bpf_ktime_get_ns() - info->pool_sync_enter;
+    return 0;
+}
+
+/*
+ * Collect data to know how much we're being throttled / delayed. If we're
+ * throttled on every tx we could hit these probes a lot (burning CPU), so we
+ * avoid using syntactic sugar in this section.
```

Comment:
What syntactic sugar is this referring to?

Reply:
I wish I had some syntactic sugar.  
The spirit of the dtrace comment, warning about not overburdening this probe seemed worth keeping.